File: nptools/predict/predict_ncnn.cpp
// Open-set classification with an exported ncnn model — reference C++ client.
//
// Mirrors nptools/predict/predict_ncnn.py. Reads openset.ncnn.param/.bin plus the class-map file, and for
// each image prints the detected class or "unknown".
//
// Build (needs ncnn and OpenCV development packages):
//
//   g++ -O2 -std=c++17 predict_ncnn.cpp -o predict_ncnn \
//       $(pkg-config --cflags --libs opencv4) -lncnn -fopenmp
//
//   # or, if ncnn is installed somewhere non-standard:
//   g++ -O2 -std=c++17 predict_ncnn.cpp -o predict_ncnn \
//       -I/path/to/ncnn/include/ncnn -L/path/to/ncnn/lib -lncnn \
//       $(pkg-config --cflags --libs opencv4) -fopenmp
//
// Run:
//   ./predict_ncnn <run>/openset.ncnn.param <run>/openset.ncnn.bin \
//                  posmlvx/class_84.txt photo.jpg [img_size]
//
// PREPROCESSING CONTRACT — every line of it matters, and none of it fails loudly if wrong:
//
//   * BGR, i.e. cv::imread's own order. Do NOT cvtColor to RGB: the model is trained on BGR.
//   * squash resize to img_size x img_size, ignoring aspect ratio (training used
//     --crop-mode=squash), with cv::INTER_AREA. INTER_AREA averages every source pixel;
//     INTER_LINEAR (OpenCV's default, and what ncnn's own from_pixels_resize does) reads a fixed
//     2x2 kernel and aliases badly on a large downscale, which shifts the distances.
//   * mean/std 0.5 on 0..1 == substract_mean_normalize({127.5}, {1/127.5}) on 0..255 pixels.
//
// This assumes the --no-argmin export: the graph emits per-class `dists` (out0) and `margins`
// (out1) and the decision is made here. That is the variant a stock ncnn build can run; the
// in-graph-argmin variant needs layers ncnn does not ship (ArgMin, Gather).
//
// The threshold is NOT needed here: it is already baked into `margins` (margins = dists -
// threshold), so the client only checks the sign at the nearest class. That keeps this code in
// lockstep with whatever was baked into the graph.

#include <algorithm>
#include <cstdio>
#include <fstream>
#include <string>
#include <vector>

#include <net.h>
#include <opencv2/imgcodecs.hpp>
#include <opencv2/imgproc.hpp>

namespace {

// One class name per line, blank lines skipped. This is the SAME file passed to openset.py, whose
// order is the exported column order — the sidecar's "class_names" is exactly these lines.
std::vector<std::string> read_class_names(const std::string& path)
{
    std::vector<std::string> names;
    std::ifstream f(path);
    if (!f) {
        fprintf(stderr, "cannot open class-map %s\n", path.c_str());
        return names;
    }
    std::string line;
    while (std::getline(f, line)) {
        while (!line.empty() && (line.back() == '\r' || line.back() == '\n' ||
                                 line.back() == ' ' || line.back() == '\t'))
            line.pop_back();       // tolerate CRLF and trailing blanks
        if (!line.empty()) names.push_back(line);
    }
    return names;
}

struct Prediction {
    int best = -1;
    float dist = 0.f;
    float margin = 0.f;
    bool unknown = true;
};

Prediction predict(ncnn::Net& net, const cv::Mat& bgr, int img_size)
{
    cv::Mat resized;
    // squash to a square, INTER_AREA — see the contract above
    cv::resize(bgr, resized, cv::Size(img_size, img_size), 0, 0, cv::INTER_AREA);

    ncnn::Mat in = ncnn::Mat::from_pixels(resized.data, ncnn::Mat::PIXEL_BGR,
                                          resized.cols, resized.rows);
    // (x/255 - 0.5) / 0.5  ==  (x - 127.5) * (1/127.5)
    const float mean_vals[3] = {127.5f, 127.5f, 127.5f};
    const float norm_vals[3] = {1.f / 127.5f, 1.f / 127.5f, 1.f / 127.5f};
    in.substract_mean_normalize(mean_vals, norm_vals);

    ncnn::Extractor ex = net.create_extractor();
    ex.input("in0", in);

    ncnn::Mat dists, margins;
    if (ex.extract("out0", dists) != 0 || ex.extract("out1", margins) != 0) {
        fprintf(stderr, "extract failed — was this exported with --no-argmin?\n");
        return {};
    }

    Prediction p;
    const int c = dists.w;                       // [C] per-class cosine distance
    const float* d = dists;
    const float* m = margins;
    p.best = 0;
    for (int i = 1; i < c; i++)
        if (d[i] < d[p.best]) p.best = i;        // argmin on the client
    p.dist = d[p.best];
    p.margin = m[p.best];
    p.unknown = p.margin > 0.f;                  // threshold is already inside margins
    return p;
}

}  // namespace

int main(int argc, char** argv)
{
    if (argc < 5) {
        fprintf(stderr,
                "usage: %s openset.ncnn.param openset.ncnn.bin class_map.txt image [img_size]\n",
                argv[0]);
        return 1;
    }
    const char* param_path = argv[1];
    const char* bin_path = argv[2];
    const char* class_map = argv[3];
    const char* image_path = argv[4];
    // Not discoverable from the .param (pnnx emits Input without dims), so it is passed in. It must
    // match the img_size in the sidecar .meta.json — a mismatch does not error, it just degrades.
    const int img_size = (argc > 5) ? atoi(argv[5]) : 128;

    std::vector<std::string> names = read_class_names(class_map);
    if (names.empty()) return 1;

    ncnn::Net net;
    // Defaults are fine for this graph; set these explicitly so behaviour does not depend on how
    // the library was built. Vulkan stays off: the head is tiny and a GPU upload costs more.
    net.opt.use_vulkan_compute = false;
    net.opt.num_threads = 4;
    if (net.load_param(param_path) != 0) {
        fprintf(stderr, "failed to load %s\n", param_path);
        return 1;
    }
    if (net.load_model(bin_path) != 0) {
        fprintf(stderr, "failed to load %s\n", bin_path);
        return 1;
    }

    cv::Mat bgr = cv::imread(image_path, cv::IMREAD_COLOR);   // BGR, alpha dropped
    if (bgr.empty()) {
        fprintf(stderr, "could not decode %s\n", image_path);
        return 1;
    }

    Prediction p = predict(net, bgr, img_size);
    if (p.best < 0) return 1;

    // Guard against a class-map that does not match the model: the exported column count is the
    // class-map line count, so a mismatch means labels would be silently shifted.
    if (p.best >= static_cast<int>(names.size())) {
        fprintf(stderr,
                "class-map has %zu entries but the model emitted index %d — wrong class-map for "
                "this model; labels would be wrong\n",
                names.size(), p.best);
        return 1;
    }

    printf("%s\n", image_path);
    printf("  detected: %s   dist=%.4f  margin=%+.4f\n",
           p.unknown ? "unknown" : names[p.best].c_str(), p.dist, p.margin);
    if (p.unknown)
        printf("  (nearest known was %s)\n", names[p.best].c_str());
    return 0;
}

File: nptools/predict/predict_ncnn.py
"""Classify images with an exported open-set ncnn model — the reference client.

This is what a device-side client has to reproduce, in the smallest form that still gets every
detail of the preprocessing contract right. Those details are where "correct in PyTorch, wrong in
ncnn" comes from, and none of them fail loudly:

  * **BGR**, i.e. cv2.imread's own order — do NOT cvtColor to RGB. The model is trained on BGR.
  * **squash** resize to img_size x img_size, ignoring aspect ratio (training used
    --crop-mode=squash), with **INTER_AREA**: it averages every source pixel, while cv2's default
    INTER_LINEAR reads a fixed 2x2 kernel and aliases badly on a large downscale.
  * mean/std 0.5 on 0..1, which on 0..255 pixels is substract_mean_normalize([127.5]*3, [1/127.5]*3).

Everything above is read from the sidecar .meta.json rather than hardcoded, and asserted where a
mismatch would otherwise be silent.

Assumes the --no-argmin export: the graph emits per-class `dists` and `margins` vectors (blobs
out0/out1) and the decision is made here. That is the variant a stock ncnn wheel can run.

Usage:
    python -m nptools.predict.predict_ncnn --model <run>/openset.ncnn --image photo.jpg
    python -m nptools.predict.predict_ncnn --model <run>/openset.ncnn --dir <data>/val --top 3
"""
import argparse
import glob
import json
import os
import sys

import cv2
import ncnn
import numpy as np

_IMG_EXTS = ('.jpg', '.jpeg', '.png', '.bmp', '.webp')


def load_meta(model_stem: str) -> dict:
    """Read the sidecar written at export time, and check it describes a model we can run."""
    for cand in (f'{model_stem}.pt.meta.json', f'{model_stem}.onnx.meta.json',
                 f'{model_stem}.meta.json'):
        if os.path.isfile(cand):
            meta = json.load(open(cand))
            break
    else:
        sys.exit(f'no sidecar .meta.json next to {model_stem}; it carries the class order and the '
                 f'preprocessing contract, and nothing works without it')
    if not meta.get('no_argmin', False):
        sys.exit('this model was exported WITHOUT --no-argmin, so the decision is inside the graph '
                 '(ArgMin + gather) and a stock ncnn wheel cannot run it. Re-export with '
                 '--no-argmin, or use a custom ncnn layer.')
    if meta.get('channel_order', 'rgb') != 'bgr':
        sys.exit(f"sidecar says channel_order={meta.get('channel_order')!r}; this client feeds BGR. "
                 f"An RGB-era model needs re-exporting (see openset.md).")
    return meta


def preprocess(path: str, meta: dict) -> ncnn.Mat:
    """Decode and normalize one image exactly as the prototypes were built."""
    bgr = cv2.imread(path, cv2.IMREAD_COLOR)          # BGR, alpha dropped
    if bgr is None:
        raise OSError(f'could not decode image: {path}')
    size = int(meta['img_size'])
    interp = cv2.INTER_AREA if meta.get('resize', 'area') == 'area' else cv2.INTER_LINEAR
    bgr = cv2.resize(bgr, (size, size), interpolation=interp)   # squash, not letterbox
    mat = ncnn.Mat.from_pixels(bgr, ncnn.Mat.PixelType.PIXEL_BGR, size, size)
    # mean/std are expressed on 0..1; ncnn works on 0..255 pixels, hence the 255x rescale:
    #   (x/255 - m) / s  ==  (x - 255m) * (1 / (255s))
    mean = [255.0 * v for v in meta['mean']]
    norm = [1.0 / (255.0 * v) for v in meta['std']]
    mat.substract_mean_normalize(mean, norm)
    return mat


def predict(net: ncnn.Net, mat: ncnn.Mat, meta: dict, top: int = 1) -> tuple:
    """Return (label, best_dist, ranked) — label is 'unknown' when the nearest prototype is too far."""
    with net.create_extractor() as ex:
        ex.input('in0', mat)
        dists = np.array(ex.extract('out0')[1])       # [C] cosine distance to every prototype
        margins = np.array(ex.extract('out1')[1])     # [C] dists - threshold; >0 means reject
    names = meta['class_names']
    order = np.argsort(dists)[:max(1, top)]
    best = int(order[0])
    # margins already carries the threshold, so the client only checks its sign. Comparing
    # dists[best] > meta['threshold'] is equivalent; margins keeps the two in lockstep.
    label = 'unknown' if float(margins[best]) > 0 else names[best]
    ranked = [(names[i], float(dists[i]), float(margins[i])) for i in order]
    return label, float(dists[best]), ranked


def main():
    parser = argparse.ArgumentParser(description=__doc__,
                                     formatter_class=argparse.RawDescriptionHelpFormatter)
    parser.add_argument('--model', required=True, metavar='STEM',
                        help='path stem of the ncnn pair, e.g. <run>/openset.ncnn '
                             '(reads STEM.param and STEM.bin)')
    parser.add_argument('--image', default='', help='one image to classify')
    parser.add_argument('--dir', default='', help='directory to scan recursively instead')
    parser.add_argument('--top', type=int, default=1, help='also show the N nearest classes')
    args = parser.parse_args()

    stem = args.model[:-len('.param')] if args.model.endswith('.param') else args.model
    stem = stem[:-len('.bin')] if stem.endswith('.bin') else stem
    param, binf = f'{stem}.param', f'{stem}.bin'
    for f in (param, binf):
        if not os.path.isfile(f):
            sys.exit(f'missing {f}')
    # strip the trailing '.ncnn' to find the sidecar, which is named after the exported model
    meta = load_meta(stem[:-len('.ncnn')] if stem.endswith('.ncnn') else stem)

    if args.image:
        paths = [args.image]
    elif args.dir:
        paths = sorted(p for p in glob.glob(os.path.join(args.dir, '**', '*'), recursive=True)
                       if os.path.isfile(p) and p.lower().endswith(_IMG_EXTS))
    else:
        sys.exit('need --image or --dir')

    empty = set(meta.get('empty_classes', []))
    print(f'model      : {param}')
    print(f'preprocess : {meta["img_size"]}x{meta["img_size"]} {meta["channel_order"].upper()}, '
          f'{meta.get("resize", "area")} resize, mean={tuple(meta["mean"])} std={tuple(meta["std"])}')
    print(f'classes    : {len(meta["class_names"])}'
          + (f' ({len(empty)} with no training data)' if empty else '')
          + f'   reject beyond dist {meta["threshold"]:.4f}'
          + (f' ({meta["threshold_deg"]:g} deg)' if 'threshold_deg' in meta else '') + '\n')

    net = ncnn.Net()
    net.load_param(param)
    net.load_model(binf)
    n_unknown = 0
    try:
        for p in paths:
            try:
                label, dist, ranked = predict(net, preprocess(p, meta), meta, args.top)
            except OSError as e:
                print(f'{p}\n  skipped: {e}')
                continue
            n_unknown += label == 'unknown'
            note = '  [no training data for this class]' if label in empty else ''
            print(f'{p}\n  detected: {label}   dist={dist:.4f}{note}')
            if args.top > 1:
                for name, d, m in ranked:
                    print(f'    {name:14s} dist={d:.4f}  margin={m:+.4f}'
                          + ('  <- rejected' if m > 0 else ''))
    finally:
        del net           # release the ncnn allocator before exit
    if len(paths) > 1:
        print(f'\n{len(paths)} images, {n_unknown} rejected as unknown')


if __name__ == '__main__':
    main()
